src/model/SplineSegment.cpp

```diff
@@ -0,0 +1,64 @@
+#include "SplineSegment.h"
+
+#include <cmath>
+
+SplineSegment::SplineSegment(const Point& p, const Point& q):
+        firstKnot(p), firstControlPoint(p), secondKnot(q), secondControlPoint(q) {}
+
+SplineSegment::SplineSegment(const Point& p, const Point& fp, const Point& sp, const Point& q) {
+    firstKnot = p;
+    firstControlPoint = fp;
+    secondControlPoint = sp;
+    secondKnot = q;
+}
+
+void SplineSegment::draw(cairo_t* cr) const {
+    cairo_move_to(cr, firstKnot.x, firstKnot.y);
+    cairo_curve_to(cr, firstControlPoint.x, firstControlPoint.y, secondControlPoint.x, secondControlPoint.y,
+                   secondKnot.x, secondKnot.y);
+}
+
+auto SplineSegment::toPointSequence() const -> std::list<Point> {
+    auto result = toPointSequenceImpl();
+    result.emplace_back(secondKnot);
+    return result;
+}
+
+auto SplineSegment::toPointSequenceImpl() const -> std::list<Point> {
+    if (isFlatEnough()) {
+        return {firstKnot};
+    }
+    auto const& childSegments = subdivide(0.5);
+    auto left_points = childSegments.first.toPointSequenceImpl();
+    left_points.splice(end(left_points), childSegments.second.toPointSequenceImpl());
+    return left_points;
+}
+
+auto SplineSegment::subdivide(float t) const -> std::pair<SplineSegment, SplineSegment> {
+
+    Point b0 = SplineSegment::linearInterpolate(firstKnot, firstControlPoint, t);  // Same as evaluating a Bezier
+    Point b1 = SplineSegment::linearInterpolate(firstControlPoint, secondControlPoint, t);
+    Point b2 = SplineSegment::linearInterpolate(secondControlPoint, secondKnot, t);
+
+    Point c0 = SplineSegment::linearInterpolate(b0, b1, t);
+    Point c1 = SplineSegment::linearInterpolate(b1, b2, t);
+
+    Point d0 = SplineSegment::linearInterpolate(c0, c1, t);  // This would be the interpolated point
+
+    SplineSegment firstPart = SplineSegment(firstKnot, b0, c0, d0);    // first point of each step
+    SplineSegment secondPart = SplineSegment(d0, c1, b2, secondKnot);  // last point of each step
+    return std::make_pair(firstPart, secondPart);
+}
+
+auto SplineSegment::linearInterpolate(const Point& p, const Point& q, float t) -> Point {
+    return Point(p.x * (1 - t) + q.x * t, p.y * (1 - t) + q.y * t);
+}
+
+auto SplineSegment::isFlatEnough() const -> bool {
+    double eps = 1.001;
```

Comment:
In practice an "average" spline segments will now be converted to a list of roughly 70 points. I hope this is fine.

Reply:
Yes, that's fine, the best would be to make this a preference, but until this is not the case, we should make a good compromise between usability, pleasant appearance, and performance / energy consumption. Otherwise we get either tickets for "high energy consumption" or a bad "look and feel".